src/tests/ftest/util/agent_utils.py

```diff
@@ -144,59 +110,59 @@ def run_agent(basepath, server_list, client_list=None):
 
             match = re.findall(pattern, output)
             if len(match) > 0:
-                print("<AGENT> agent started on node {} in {} "
-                      "seconds".format(client, time.time() - start_time))
+                print(
+                    "<AGENT> agent started on node {} in {} seconds".format(
+                        client, time.time() - start_time))
                 break
 
     return sessions
-# pylint: enable=too-many-locals
 
 
 def stop_agent(sessions, client_list=None):
-    """
-    This should kill ssh and the agent.  This is temporary; presuming the
-    agent will deamonize at somepoint and can be started killed more
-    appropriately.
+    """Kill ssh and the agent.
 
-    client_list -- kill daos_agent on these hosts
-    sessions    -- set of subprocess sessions returned by run_agent()
-    """
+    This is temporary; presuming the agent will deamonize at somepoint and can
+    be started killed more appropriately.
+
+    Args:
+        sessions (dict): set of subprocess sessions returned by run_agent()
+        client_list (list, optional): lists of hosts running the daos agent.
+            Defaults to None.
+
+    Raises:
+        AgentFailed: if the daos agents failed to stop
 
+    """
     # if empty client list, 'self' is effectively client
     if client_list is None:
         client_list = [socket.gethostname().split('.', 1)[0]]
 
-    # this kills the agent
-    for client in client_list:
-        cmd = "pkill daos_agent"
-        resp = subprocess.call(["ssh", client, cmd])
+    # Kill the agents processes
+    clush(client_list, "pkill daos_agent", None, False)
 
+    # Kill any processes running in the sessions
     for client in sessions:
         if sessions[client].poll() is None:
             sessions[client].kill()
         sessions[client].wait()
 
-    # check to make sure it's dead
+    # Check to make sure all the daos agents are dead
+    # pgrep exit status:
+    #   0 - One or more processes matched the criteria.
+    #   1 - No processes matched.
+    #   2 - Syntax error in the command line.
+    #   3 - Fatal error: out of memory etc.
     time.sleep(5)
-    found_hosts = []
-    for host in client_list:
-        proc = subprocess.Popen(["ssh", host,
-                                 "pgrep 'daos_agent'"],
-                                stdout=subprocess.PIPE)
-        stdout = proc.communicate()[0]
-        resp = proc.wait()
-        if resp == 0:
-            found_hosts.append(host)
-
-    if found_hosts:
-        raise AgentFailed("Attempt to kill daos_agent processes {} on hosts "
-                          "{} was unsuccessful"
-                          .format(', '.join(stdout.splitlines()), found_hosts))
+    result = clush(client_list, "pgrep 'daos_agent'", None, False)
+    if result.exit_status != 1:
+        raise AgentFailed(
+            "DAOS agent processes detected after attempted stop:\n{}".format(
+                result.stdout))
```

Comment:
Can we easily report which hosts they were on, to help somebody debugging with which logs to look at?

Reply:
Yes.  The new method returns a dictionary of return codes and the nodes with those return codes.